zeppelin-server/src/main/java/org/apache/zeppelin/service/NotebookService.java

```diff
@@ -340,27 +344,63 @@ public void runAllParagraphs(String noteId,
       return;
     }
 
-    note.setRunning(true);
-    try {
-      for (Map<String, Object> raw : paragraphs) {
-        String paragraphId = (String) raw.get("id");
-        if (paragraphId == null) {
-          continue;
-        }
-        String text = (String) raw.get("paragraph");
-        String title = (String) raw.get("title");
-        Map<String, Object> params = (Map<String, Object>) raw.get("params");
-        Map<String, Object> config = (Map<String, Object>) raw.get("config");
-
-        if (!runParagraph(noteId, paragraphId, title, text, params, config, false, true,
-                context, callback)) {
-          // stop execution when one paragraph fails.
-          break;
+    synchronized (note) {
+      if (note.isRunning()) {
+        LOGGER.warn("Can't run note because it already is running");
+        return;
+      }
+      note.setRunning(true);
+    }
+
+    noteExecutor.execute(() -> {
+      try {
+        for (Map<String, Object> raw : paragraphs) {
+          String paragraphId = (String) raw.get("id");
+          if (paragraphId == null) {
+            continue;
+          }
+          String text = (String) raw.get("paragraph");
+          String title = (String) raw.get("title");
+          Map<String, Object> params = (Map<String, Object>) raw.get("params");
+          Map<String, Object> config = (Map<String, Object>) raw.get("config");
+
+          if (!runParagraph(noteId, paragraphId, title, text, params, config, false, true,
+              context, callback)) {
+            // stop execution when one paragraph fails.
+            break;
+          }
         }
+      } catch (IOException e) {
+        LOGGER.error("Can't run paragraph", e);
```

Comment:
does this get reported to the user in the frontend?

Reply:
I think not worth it. Since if something goes wrong during the start of a paragraph, then an error message will be displayed in the paragraph.